src/tools/ceph_immutable_object_cache/CacheSession.cc

```diff
@@ -0,0 +1,109 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/debug.h"
+#include "common/ceph_context.h"
+#include "CacheSession.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_immutable_obj_cache
+#undef dout_prefix
+#define dout_prefix *_dout << "ceph::cache::CacheSession: " << this << " " \
+                           << __func__ << ": "
+
+
+namespace ceph {
+namespace immutable_obj_cache {
+
+CacheSession::CacheSession(uint64_t session_id, boost::asio::io_service& io_service, ProcessMsg processmsg, CephContext* cct)
+    : m_session_id(session_id), m_dm_socket(io_service), process_msg(processmsg), cct(cct)
+    {}
+
+CacheSession::~CacheSession() {
+  close();
+}
+
+stream_protocol::socket& CacheSession::socket() {
+  return m_dm_socket;
+}
+
+void CacheSession::close() {
+  if(m_dm_socket.is_open()) {
+    boost::system::error_code close_ec;
+    m_dm_socket.close(close_ec);
+    if(close_ec) {
+       ldout(cct, 20) << "close: " << close_ec.message() << dendl;
+    }
+  }
+}
+
+void CacheSession::start() {
+  handing_request();
+}
+
+void CacheSession::handing_request() {
+  boost::asio::async_read(m_dm_socket, boost::asio::buffer(m_buffer, RBDSC_MSG_LEN),
+                          boost::asio::transfer_exactly(RBDSC_MSG_LEN),
+                          boost::bind(&CacheSession::handle_read,
+                                      shared_from_this(),
+                                      boost::asio::placeholders::error,
+                                      boost::asio::placeholders::bytes_transferred));
+}
+
+void CacheSession::handle_read(const boost::system::error_code& err, size_t bytes_transferred) {
+  if (err == boost::asio::error::eof ||
+     err == boost::asio::error::connection_reset ||
+     err == boost::asio::error::operation_aborted ||
+     err == boost::asio::error::bad_descriptor) {
+    ldout(cct, 20) << "fail to handle read : " << err.message() << dendl;
+    close();
+    return;
+  }
+
+  if(err) {
+    ldout(cct, 1) << "faile to handle read: " << err.message() << dendl;
+    return;
+  }
+
+  if(bytes_transferred != RBDSC_MSG_LEN) {
+    ldout(cct, 1) << "incomplete read" << dendl;
+    return;
+  }
+
+  process_msg(m_session_id, std::string(m_buffer, bytes_transferred));
+}
+
+void CacheSession::handle_write(const boost::system::error_code& error, size_t bytes_transferred) {
+  if (error) {
+    ldout(cct, 20) << "session: async_write fails: " << error.message() << dendl;
+    assert(0);
+  }
+
+  if(bytes_transferred != RBDSC_MSG_LEN) {
+    ldout(cct, 20) << "session : reply in-complete. "<<dendl;
+    assert(0);
+  }
+
+  boost::asio::async_read(m_dm_socket, boost::asio::buffer(m_buffer),
```

Comment:
In the current implementation, each instance of `async_read` is called by a matching `async_write`, making them essentially blocking operations -- only after an `async_write` finishes its checks (the checks in `handle_write`, before the `async_read` was called) does the corresponding `async_read` gets called. 

One way to fix this is to move the entire `async_read` call from `handle_write` to the end of `handle_read`. Doing so will decouple read and write operations at client side, thus achieving real asynchronous I/O. After a read operation was received and handled, it will start to process another pending operation instead of waiting on the write operation to actually complete.

Reply:
@usefulalgorithm thanks for sharing your ideas! I think you're concerning about the performance. As the daemon is just doing IPC with small messages, on my testing env with a poor Xeon(SNB-EP based) the daemon is able to handle ~400k IOPS, which I guess should be able to cover for most workloads?

In the improvement plan we have rough ideas to use some thread pool based [design](https://www.boost.org/doc/libs/1_63_0/doc/html/boost_asio/example/cpp03/http/server3/server.cpp), on `async_read`, we should store the message into a queue with a unique client id, then the daemon would pop message from the queue and do the processing work, on `async_write`  we just ack the right clients with reply, but this would require some extra logic to handle the message mapping/locking. We would continue looking into this and discuss with the community

Reply:
Hi @zhouyuan ,

>  I think you're concerning about the performance

Sort of... in my environment I basically just setup a minimalistic cluster via `vstart.sh` (because I'm lazy), fire up the daemon and do `fio` on the RBD image. After moving `async_read` call, performance went up from 12X Mb/s to 19X Mb/s, which is quite a significant boost. I guess ~400k IOPS is good enough for now, but still...

> we have rough ideas to use some thread pool based design

Just curious, but does this have anything to do with the `ThreadPool` construct that happened several commits ago? In my opinion `CacheController` (and `CacheServer` as well) having a thread pool to handle operations is indeed the way to go, but I'm not sure about how multi-threading should work for client side operations. 

> on async_read, we should store the message into a queue with a unique client id (...) we just ack the right clients with reply

I think in the current implementation the "matching messages with the right clients" part was already taken care of? Each client communicates with a `CacheSession`, which was kept and managed by the server. 

On the other hand, the queue might solve the problem of partial cached I/Os. Currently in my heavily modified version, when an (asynchronous, non-blocking) request arrives at server side, the requested object gets promoted; but when the other requests for the same object with different offsets arrive (this happens if your blocksize is smaller than the object size), server will reply them with `RBDSC_READ_RADOS`, making them essentially uncached I/Os. Previously I considered partial promotion as a potential solution to this issue, but now that I think of it, having a queue at server side to postpone response in order to achieve higher cached I/O rate is probably a better and easier way to do this.

Thanks!

Reply:
@usefulalgorithm In case you didn't notice, recent commit (7e6a57bd2) changed the default build to "Debug", see if switching to "Release" will bring the performance back. 
In current implementation besides the single-thread daemon, another big overhead is from the cache read(sync read) in `handle_cache_read`. async read based on workqueue or libaio should help here. I'll need some time to improve the current error handling first to make the code more solid, then try to improve the performance as next. Thanks again for the great inputs!